#53 Use single style rule

nickmoreton opened the pull request:
Ticket: https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/53

### This pr changes the way we now parse the html inline style attributes as well as some html tags.

It now uses individual style rules, rather than the complete style string, therefore we can transform inline styles to html tags to represent the style by matching only one style rule from a complete style string. It also transforms some HTML tags for better Wagtail compatibility.

The config for searching styles can be overridden by a developer in their own Wagtail settings. It's intended that a developer will be able to provide their own methods to transform the styles and tags.

We intend to provide a basic/most common set of rules in the package.

### _Example:_

```<span style="font-weight: bold; font-style: italic;">text</span>```
would be transformed to ```<b><i>text</i></b>``` when used to populate the wagtail streamfield blocks

### _Note_
There's likely more work to do when creating blocks based on the parsed body content. At that point the style attr will still be in place for us to use. Some style attrs such as float and text-align are transformed to classes. This is one piece that we will use later when building out the required blocks for the stream field. e.g. when setting image alignment in a rich text block.

Currently the style attrs are removed by bleach at the last stage.
I'm undecided on the order of where bleach should be used in the end but suspect we may need 2 bleach operations with different settings for each call.

Docs: they are slightly out of date now but I don't want to update them until this PR is merged to avoid conflicts. https://github.com/torchbox/wagtail-wordpress-import/pull/36

nickmoreton committed:
add transform styles filter

nickmoreton committed:
change settings name for debug

nickmoreton committed:
remove files not required

nickmoreton committed:
fix broken tests

nickmoreton committed:
change allowed attrs for bleach filter

nickmoreton committed:
update tests fixture

nimasmi commented on wagtail_wordpress_import/prefilters/transform_styles_filter.py:
```diff
@@ -0,0 +1,221 @@
+def normalize_style_attrs(html, options=None):
+    """
+    There are different ways that styles are formatted coming out of wordpress.
+    This normalizes them so the know what the format is for later parsing.
```
**Suggestion**: Fix a typo and reword anyway since I was suggesting.

```suggestion
    This normalizes them so we have a predictable format is for later parsing.
```

nickmoreton replied:
I'll apply a fix here

nickmoreton replied:
I ended up changing this so it's inline with the method changes.

nimasmi commented on wagtail_wordpress_import/prefilters/transform_styles_filter.py:
```diff
@@ -0,0 +1,221 @@
+    Worth noting that these styles are actually removed later on when the fix_styles
+    method is run so if they look wrong it's OK here as they are a `template`
+
+    param: `options` NOT IMPLEMENTED
```
**Question**: …but it's a non-negotiable standard parameter for a prefilter?

nickmoreton replied:
I have reworded this, I think it was misleading.

nimasmi commented on wagtail_wordpress_import/prefilters/transform_styles_filter.py:
```diff
@@ -0,0 +1,221 @@
+                for style in el.attrs["style"].split(";")
+                if style != ""
+            ]
+            el.attrs["style"] = " ".join(sorted(styles_list))
```
**Question**: We're joining with a `" "` here. Why? Why not just concatenate them without spaces for simpler parsing later?

nickmoreton replied:
I suspect it was left like this when the function was moved. It works as expected without the space.

nimasmi commented on wagtail_wordpress_import/prefilters/transform_styles_filter.py:
```diff
@@ -0,0 +1,221 @@
+        if callable(styles_mapping):
+            CONF_STYLES_MAPPING = styles_mapping()
+        else:
+            CONF_STYLES_MAPPING = styles_mapping
```
**Comment**: There's a styles/filter naming discrepancy here. Is this a hangover from other refactoring? We only use it once, so it seems okay here to keep it as `CONF_FILTER_MAPPING`, but if the renaming is important, then we should rename it in the config dict too.

nickmoreton replied:
Yes thats makes sense, I have renamed to to `TRANSFORM_STYLES_MAPPING` so the config stays the same. We'll need to update the docs on that too. I'll create a docs ticket like the tests ticket for it.

nimasmi commented on wagtail_wordpress_import/prefilters/transform_styles_filter.py:
```diff
@@ -0,0 +1,221 @@
+
+    new_tag = soup.new_tag("b")
+    new_tag.attrs["style"] = tag.attrs["style"]
+    new_tag.string = tag.text
```
**Question**: This is probably me being rusty with Beautifulsoup, but why do we read in the text attribute but then set that value as the string attribute of the replacement tag? It looks like it's legitimate to modify the string attribute as we do, https://www.crummy.com/software/BeautifulSoup/bs4/doc/#modifying-string, but I can't find mention of what `.text` is or does in those docs.

NB this question is relevant to many lines here.

nickmoreton replied:
Here we need to put back the attr as a string ready for if the next time around. Each tag could need to be parsed more than once like if there is bold and italic attrs on one tag.
.text is the inner content of the tag which is added as the content of the new tag. I did it this way because it's possible that inner content could be html.

nimasmi replied:
Okay. Reading `.text` gives you everything, whereas reading `.string` only gives you string contents, but it's safe to write HTML to `.string` anyway? I think I follow.

nimasmi commented on wagtail_wordpress_import/prefilters/transform_styles_filter.py:
```diff
@@ -0,0 +1,221 @@
+
+        for tag in tags:
+            if tag.name not in CONF_HTML_TAGS:
+                print("item.name = tag not found in HTML_TAGS")
```
**Suggestion**: Are we only at the 'log by printing things' stage of the project at the moment? Do we have any plans yet for more detailed reporting or logging?

nickmoreton replied:
This one is the only logging thats not yet output to a file. We'll need to deal with that I missed it :) I have added it to the issue for this part of the project and made a note in the ticket.

nimasmi commented on wagtail_wordpress_import/prefilters/transform_styles_filter.py:
```diff
@@ -0,0 +1,221 @@
+from django.utils.module_loading import import_string
+
+
+def normalize_style_attrs(html, options=None):
```
**Comment**: It's a bit clunky that this function takes a string, parses it with Beautifulsoup, and returns a string, only for the calling code at https://github.com/torchbox/wagtail-wordpress-import/pull/48/files#diff-ba7adf08e8b8a91567d4cac5f2059c5cfee02371cfc8aad3b80d29a1046d931aR55 to parse the output a second time. Would it be possible for this function to take and return a soup object?

nickmoreton replied:
I see what you mean. Will change it to a soup object all through.

nimasmi replied:
Hmm, the way you've done it, it's now clunky that this function takes a string, and returns a soup object. I'd prefer the parsing to be done elsewhere (line 55), and then this take a soup object only, as well as returning one.

Beautifulsoup doesn't have any type stubs, so I'm not sure how we'd annotate this. Can you make it clear what the argument and return types are in the docstring?

nimasmi reviewed (commented):
It's a bit daunting to see this with renamed and moved files as my first MR. Thanks for your patience.

I am additionally concerned that there is no direct unit testing of the input and output of the `normalize_style_attrs` function and the `filter_transform_inline_styles_to_tags` functions. Have I missed that, or is it only tested in the one test case as we discussed?

nickmoreton committed:
code review updates

incudes help text, conf names and soup objects

nickmoreton committed:
removed test_functions.py

It was failing becuse of code review changes but is to be added back in in the testing tickets

nickmoreton committed:
join style without a space

nickmoreton committed:
update normalize_style_attrs() usage

it now takes a soup object in and returns modified soup object

nimasmi reviewed (approved):
With the exception of the deleted test file (which seemed a reasonable start, and probably could have been left in with slight tweaks), this is approved.

Is that deleted test file captured in the testing tickets, now?
